Adds IPCs

Ziiro opened the pull request:
:cl: Ziiro
adds: Adds IPCs as a roundstart selectable species. (Posibrain code by Flattest)
/:cl:

Special thanks to:
Traveling Merchant for Shellguard
BRAINOS of Aurora for: Bishop Cyberkinetics 2.0, Xion Manufacturing Group 2.0, Hephaestus Industries 2.0, Zeng-Hu Pharmaceuticals
Hippiestation for the Greyscaled Morpheus Cyberkinetics
...Artists for the other sets I can't credit! Because I have no idea where they came from. If you are one of those people, please message me on Oracle discord or leave a message here!

Under the hood:
* Chem processing restrictions based on Synthetic/organic tags
* Protected_species flag for game mode
* https://github.com/tgstation/tgstation/pull/31202
* Robotic Surgery (Thanks, Flattest)
* NOSCAN, REVIVESBYHEALING, NOCHANGELING, NOHUSK, ROBOTIC_LIMBS, NOMOUTH species traits

Features:
* IPCs are properly robotic - repair with cable coil and a welder. (No bones, no blood.)
* All organs (Except posibrain) can be printed from the mech fab.
* Robotic emotes are attached to the robotic voicebox (Some are clown-restricted)
* Re-attach limbs, anytime, anywhere!
* Unique EMP effects!
* Take more damage from nearly all sources (and can be dismembered by blunt objects if the force is high enough)
* Recharge at APCs, or drink Plasma/Stable Plasma for nutrition.
* Simple robotic surgery using tools! (Screwdriver, crowbar, multi-tool.)
* They do not breathe, aren't affected by diseases/radiation, 
* They cannot be changelings, nor can they be absorbed by them.
* They cannot be cloned or have DNA modified
* They do not process most chems (and have two unique ones. Liquid Solder for brain damage, and System Cleaner to function like charcoal)

---
Old Original post:

Obviously this isn't merge ready just yet, but I thought I'd throw this up here to reference for people to help me work on my shitty code as I go. (And as a resource if I fuck off and someone else takes up the mantle)

Current bugs/issues as of 10/26/2017:
- Screens/antennas use the body mutant_color, which looks shitty.
- They still use organic meat brains (Need to figure out how to turn a posibrain into an organ)
- NOLIVER makes them have no liver, and not process chems. I plan to make a custom organ for this that functions as one.
- Uses NOHUNGER, but this is a bandaid. I'd prefer to have a cell that lets them recharge in cyborg rechargers
- Sometimes the left leg looks like the wrong augment set. I don't know. **Fuck.**
-The scream is the human scream
-Lacks the full suite of emotes (*beep *yes *no *buzz)
- No random namelist (If I can figure out how to hook the posibrain namelist into it, that would work)
- Surgery options are organic surgery (/tg/ has a solution for this we'll get later)
- Bare minimum of EMP interaction. The stuff that is EMP-able comes from already built in stuff for the robotic limbs/eyes.

Ziiro committed:
Initial Commit

Ziiro committed:
Adds basic bodypart icons. Tweaks some numbers. Adds damage overlays.

Ziiro committed:
Snowflakes bodypart code (and ipc.dm updates)

Ziiro committed:
Accessories

Ziiro committed:
Fixes a couple sentences

Ziiro commented:
>Sometimes the left leg looks like the wrong augment set. I don't know. Fuck.

Okay so this is something that gets fixed if my color gets changed by a magic mirror.  _What._

![2017-10-26_19-09-05](https://user-images.githubusercontent.com/26494679/32083613-4528b490-ba81-11e7-90aa-2202b0f936de.gif)

imsxz commented:
To those who have access to the pending comments I made, they're incorrect and I'm a doofus. Delet them please, idk how to from mobile.

Ziiro committed:
Fixeds prefs saving. Adds organs with EMP interactions.

Birdtalon commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Port IPC](https://trello.com/c/XkCMBJBc/27-port-ipc)

Joseph-Case commented:
Quick fix for your CL
:cl: Zirro
add: Eveyone's favourite EMP Magnet
/:cl:


In terms of the PR, Looks good, I can finally ~~meme~~ play a "mechanically" different race
*fucking kill me for that pun*

Ziiro committed:
Adds isipcs, fixes sneezing, coughing, and screaming. Adds emotes.

Ziiro committed:
IPC deathgasp, random mutant_bodyparts.

Ziiro commented:
My to do list, in order of priority to become "playable".

- Use Rechargers for Nutrition: (Somewhere in here, probably?): https://github.com/OracleStation/OracleStation/blob/7056b384106ef0e50cf0c8bcf9be802df4e2eef5/code/game/machinery/rechargestation.dm#L113
- Port Robo Surgery from /tg/ (Waiting on flattest to finishing surgery before I consider touching this)
- Add way to fix IPC organ damage (Probably attached to above)
- Use Posibrain as the brain in chest(Might have to put in mutant_brain from further downstream? I also can't wrap my head around how to do this.)
- Revival not requiring defibrillator
- Remove and and all chem processing except for Holy Water. (Don't do drugs, beep boop.)
- Re-attaching limbs needs a progress bar, it's currently instantaneous ( Somewhere in here:) https://github.com/OracleStation/OracleStation/blob/962d8976110eb273aa9a086ce891963d09b1aebd/code/modules/surgery/bodyparts/bodyparts.dm#L63 
- Balancing damage taken/other factors (Limbs currently only fly off when hit by sharp weapons. _But boy do they fly!_)
- On hit effects effect needs to be changed to spark
- Gibbing needs to be changed to Robotic
- Screen/Antenna colorable (no precedent in /tg/ code for mutant_bodyparts having unique coloring)
- Add Namelists (Using posibrain namelists)
- EMPing special effects (Sparking, screaming.)
- **BONUS** Emag effects.
- **BONUS** Give Science the ability to produce IPC Chassis

FlattestGuitar commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -201,3 +201,142 @@
 		T.Entered(src)
 //Ayy lmao
+
+// IPC Emotes.
```
Most of these are made for humans only in your code, but there are also versions for cyborgs, most of which are just duplicated here. Would be cleaner if you made the versions for cyborgs function for both them and IPCs.

Ziiro replied:
Unfortunately, without the ability to define a mob past `/mob/living/carbon/human` (for example `/mob/living/carbon/human/species/ipc`) it makes me use the hacky solution that imsxz used for *squish. I tested it, and humans and other organics can't use the IPC emotes. 

I'd likely need to grab [this](https://github.com/tgstation/tgstation/blob/aac6b076c704ddafa408f4b67edf5d1a6d41be02/code/modules/mob/living/carbon/human/human.dm#L912) from further downstream.

Once we get that would probably be as easy as putting it here: https://github.com/OracleStation/OracleStation/blob/master/code/modules/mob/living/silicon/robot/emote.dm#L2

And I can erase this entire section.

Ziiro replied:
Update: I attempted the above, it did not work. Back to the drawing board.

FlattestGuitar commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -201,3 +201,142 @@
+	if(isipc(user))
+		return TRUE
+
+ // MAKE THESE LIMITED TO CLOWNS ONLY. SOMEHOW.
```
Check for `user.mind.assigned_role == "Clown"` for these.

Assuming that's the reason you commented them out. 🤡

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,40 @@
+	coldmod = 2 // Don't put your computer in the freezer.
+	burnmod = 1.5 // Overcoming augmented damage reduction - 1.5 puts it on par with unaugged organics.
+	brutemod = 1.5
+	species_traits = list(NOBREATH,NOBLOOD,RADIMMUNE,VIRUSIMMUNE,EASYDISMEMBER,EASYLIMBATTACHMENT,NOPAIN,NO_BONES,NOCLONE,NOTRANSSTING,MUTCOLORS,TOXINLOVER) // NOHUNGER is temporary. A liver is needed to process chems (LIKE HOLY WATER)
```
You'll probably need to code a special cybernetic liver for them to be able to process chems properly. Also, what's with `TOXINLOVER` being on in here?

Ziiro replied:
`TOXINLOVER` just makes them immune to toxins for the moment, because healing chems don't heal robotic limbs anyway.

I'll redo it when I add the substance processor.

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,40 @@
+	species_traits = list(NOBREATH,NOBLOOD,RADIMMUNE,VIRUSIMMUNE,EASYDISMEMBER,EASYLIMBATTACHMENT,NOPAIN,NO_BONES,NOCLONE,NOTRANSSTING,MUTCOLORS,TOXINLOVER) // NOHUNGER is temporary. A liver is needed to process chems (LIKE HOLY WATER)
+	mutant_bodyparts = list("ipc_screen", "ipc_antenna")
+	default_features = list("mcolor" = "#7D7D7D", "ipc_screen" = "Static", "ipc_antenna" = "None")
+	meat = null
```
For the hilarity value I'd make cybersteaks, tbh.

Ziiro replied:
Actually, how about metal sheets?

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,40 @@
+		Z.change_bodypart_status(BODYPART_ROBOTIC, FALSE, TRUE) // Makes all Bodyparts robotic
+		Z.icon = 'icons/mob/augmentation/ipc.dmi' // Overrides the augmentation icons
+/*
+/datum/species/ipc/proc/emp_act // Other EMP effects are handled in the individual organs. I'd like the IPC to spark though.
```
Define this proc first for `/datum/species`, only make it do stuff here.

FlattestGuitar commented on code/modules/mob/living/emote.dm:
```diff
@@ -496,7 +497,7 @@
 		else
 			sound_to_play = H.dna.species.male_scream_sound
-	if(issilicon(user))
+	if(issilicon(user) || isipc(user))
```
The above checks link specifically to the species scream sounds, just put the right file on `datum/species/ipc`, not here.

FlattestGuitar commented on code/modules/mob/living/emote.dm:
```diff
@@ -519,7 +520,7 @@
 		else
 			sound_to_play = H.dna.species.male_cough_sound
-	if(iscyborg(user))
+	if(iscyborg(user) || isipc(user))
```
As the above comment, define the sounds for `/datum/species/ipc`.

FlattestGuitar commented on code/modules/mob/living/emote.dm:
```diff
@@ -542,7 +543,7 @@
 		else
 			sound_to_play = H.dna.species.male_sneeze_sound
-	if(iscyborg(user))
+	if(iscyborg(user) || isipc(user))
```
As the above comment, define the sounds for `/datum/species/ipc`.

FlattestGuitar commented on code/modules/surgery/bodyparts/bodyparts.dm:
```diff
@@ -361,7 +361,7 @@
 		should_draw_gender = FALSE
 
-	if(status == BODYPART_ORGANIC)
+	if(status == BODYPART_ORGANIC || icon == 'icons/mob/augmentation/ipc.dmi' ) // S.N.O.W.F.L.A.K.E.
```
Oh. God.

Two ways to go about this:

a) just roll with regular robotic bodypart sprites. We'd want those to be the same thing as regular cybernetic limbs everyone can add to themselves anyway, so it will help us avoid confusion and the IPC sprites are shit anyway..

b) Add a specific variable on the species datum that links to the icon file that should be checked when you're forging the icons.

Ziiro replied:
Okay, this is going to take some explanation.

I had to make IPCs spawn with a full set of augmented limbs in order to have them properly inherit the `BODYPART_ROBOTIC` flag, and everything attached to that. Repair, EMP effects, sounds, limb properties, etc.

Augmented limbs (any their few flavors of Grey/Standard/Miner/Sec/Eng) ignore all inherited color. This is to prevent an augmented limb from looking flesh colored. It makes sense in the case of the rest of the augmented limbs, obviously, but not for IPCs. This solution I've used here is _fucking terrible_ and I am ashamed of it, but I was unable to figure out a better solution to getting augmented limbs to render with color.

In regards to the two ways:

a) MUH SCREENS AND ANTENNAS - the standard `BODYPART_ROBOTIC` limbs are for augmentation, not full robits. Some of those can be seen [here](https://github.com/tgstation/tgstation/pull/28977). They're still very human shaped. 

b) This is already done in their `on_species_gain` shenanigans. `Z.icon = 'icons/mob/augmentation/ipc.dmi'`

If I can figure out a way to make their icons render with color while retaining `BODYPART_ROBOTIC`, this goes away.

FlattestGuitar commented on code/modules/surgery/organs/ears.dm:
```diff
@@ -93,4 +93,25 @@
+		if(1)
+			owner.Jitter(120)
+			owner.Dizzy(120)
+			deaf = 1
```
`deaf` ticks down every two seconds. Make it a higher value to make it meaningful. Maybe somewhere around 20 seconds?

FlattestGuitar commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -86,3 +86,21 @@
 		if(2)
 			damage+=50
+
+/obj/item/organ/liver/ipc  // It's a liver because a liver is needed to process chems (Like holywater) but it doesn't make sense for a robot to have a liver.
```
Not a fan of this explanation. When you take out an IPC's "Morality Core" they suddenly can't process chems?

I'd call it a Substance Processor or something, so that it makes sense.

FlattestGuitar reviewed (changes requested):
Still a long way to go. :P

Ziiro committed:
Added Whitespace Rights. Fixed Head Accessories (AND COLORABLE!). Made Honk and Sad clown-only. Removed TOXINLOVER. Added Metal Sheet when gibbed (and removed skin.) Fixed Sneezing/Screaming/Coughing. Increased robot ears deafness severity. Changed Liver.

Ziiro commented:
![dreamseeker_2017-10-28_22-06-22](https://user-images.githubusercontent.com/26494679/32140761-c88b5dea-bc31-11e7-9f35-91867524d628.png)

Hair color = Antenna Color
Eye color = Screen color.

That was way easier than I thought.

**_WHY ARE THE LEGS STILL DOING THSI!?_**

Ziiro committed:
Some renames. Fixes some numbers. Adds TOXMOD. Fixes default look.

Ziiro committed:
LOTS OF CHANGES.
Adds: random name, chassis selection, organ adjustments, balance adjustments, potential for chem filtering.
Breaks: Body colors. rip.

Xantholne commented:
Toasters when

Ziiro commented:
I can probably solve a few more of these, but there's a wall I'm going to hit soon because I only started coding last week.

Here's what I consider my list to be "ready" (though I'm sure all maintainers disagree with every part of it)

High Priority:
- Chem/drug filtering.
- Revival not requiring defib
- Use Posibrain as the Brain (In chest would be preferable, but at this point I'd accept it in the head.)
- Remove appendix
- LIMBS NEED TO BE COLORABLE. **FUCK.** HOW CAN YOU TELL THE TOASTERS APART IF THEY AREN'T COLORFUL. I have to think of a new solution.

Low Priority/Balance:
- Re-attaching limbs needs a progress bar, not instanteous.
- Robo-surgery from /tg/ (Things are going to be hilarious before this happens)
- EMP special effects (sparking)
- Effect when shot needs to be changed.

Ziiro committed:
Adds:
REVIVESBYHEALING species_trait, and using it in mob_helpers for robotic repair.
Finishes up most EMP effects. Still missing: Sparks.

Ziiro committed:
Integrates https://github.com/tgstation/tgstation/pull/31202 for mutant_brain

Ziiro committed:
Posibrain Brains added. Broken as shit currently, though!

Ziiro commented:
![2017-11-03_21-10-06](https://user-images.githubusercontent.com/26494679/32401819-67c1de08-c0dc-11e7-9b5c-d28ff7807fa2.gif)

Posibrains for brains have replaced the brain (they're in the chest!). A few issues remain/are untested, however:
- /tg/ Posibrains can't be muted, currently. Or if they can it's a testing issue.
- For some reason, they can't be put into borgs? Maybe. Again. Testing issue.
- Haven't tested removing brain through surgery and putting it back.
- Not sure if beheading actually keeps them alive or not.
- I'm not sure if an IPCs posibrain counts as dead or alive.

General issues:
- Robotic Gibs added, but the old gibs are still there too.
- Eyes appear on the disembodied head instead of screens

Remaining:
- Chassis selection code sucks. (help)
- **STILL CAN'T COLOR THE BODY.** (_PLS HELP_)
- Unique organs have no icons
- Chem filtering is still to-do

Ziiro commented:
Oh also the posibrain stuff runtimes like a motherfucker

Ziiro committed:
Reverts Para Posibrain port. Implements IPC Posibrain unique organ.

Ziiro commented:
Too many runtimes. Too much bullshit trying to use Para's system wholesale. New solution included, but they aren't real posibrains.

Going to solve other problems and come back to it.

Ziiro committed:
Changelog:
Adds some code for jamming an IPC posibrain in an MMI (sprite TBD)
Robot emotes are now connected to having a robotic voicebox
IPCs now generate steel/plasteel when the put into gibber
IPCs now spark when shot instead of bloodspattering
Minor organ text changes

Ziiro committed:
Adds processing flags for ORGANIC and SYNTHETIC.

Ziiro commented:
*NEW* FEATURES:
- Silicon emotes are accessible to anyone with a robotic voicebox
- IPCs Spark when shot (1 spark per projectile hit)
- Gibber now spits out 5 Plasteel and 10 Metal when you put an IPC in it
- Posibrain!.. kind of. Will require an MMI "Adapter" in order to be plugged in to Cyborg/AI cores. Sprites are in progress. I'm not 100% happy with this implementation.

Remaining issues before "Feature complete"

- Chassis selection code sucks. (help)
- STILL CAN'T COLOR THE BODY. (PLS HELP)
- Unique organs/Posibrain+MMI has no icons (working on it)
- Gibs aren't robotic gibs. Help.
- No code yet for limbs falling off based on exceeding their damage cap. This is necessary for balance reasons.
- Attaching limbs needs a progress bar for balance reasons.

Additional:
- REVIVESBYHEALING exists as a species flag, but currently organics have nowhere to take advantage of this. Suggestions?
- There are no Synthetic-specific chems currently, and I currently have no plans to re-make Ultra-lube/Degreaser/etc. I need to decide on a chem to fix their brain/organ damage though.

Ziiro committed:
FIXES COLOR (WITH SLIGHTLY LESS SHITTY SNOWFLAKE CODE)
Chassis system removed (until I can make mutant_body)
adds timer for self limb re-attachment

Ziiro commented:
Chassis Removed. (I'm going to need to make mutant_body eventually)
Color re-added. (I'm sorry for my code.)
Balance adjustments made to limb re-attachment
The left leg issue is gone. I don't know why.

Remaining issues:
- Meat Gibs
- Limbs don't fall off from pure damage
- Floor icons don't work properly
- Message missing on repairing a full health, conscious target
- Melee attack need sparks
- Attacks should create oil(blood) stains
- Squish noise when 'gutted' should be changed (not sure how feasible this is)
- Need to adjust some chem to cure brain damage / synthetic organ damage
- Missing Icons

Ziiro committed:
Breaks colors. Again.
Adds temporary icons to new organs.
Fixes up REVIVESBYHEALING code in item_heal_robotic a bit.

Ziiro committed:
Adds liquid solder for synthetic brain damage

Ziiro committed:
Adds var/species_gibs to species.dm to define what types of gibs the species should have
Fixes splort sound being played even if the chest is augmented

SamHPurp commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -146,3 +146,15 @@
+/obj/item/organ/brain/ipc
+	name = "IPC posibrain"
+	slot = "brain"
+	desc = "A cube of shining metal, three inches to a side and covered in shallow grooves. This particular model seems specific to IPCs, due to lack of foresight. It will need an MMI adapter to be slotted into cyborgs or AI cores."
```
If its a 3inch cube the size should be reduced.

It should probably be made to be of comparable size to regular brains, updating size & description accordingly.

SamHPurp commented on code/modules/mob/living/carbon/human/death.dm:
```diff
@@ -1,20 +1,29 @@
 /mob/living/carbon/human/gib_animation()
-	new /obj/effect/temp_visual/gib_animation(loc, "gibbed-h")
-
-/mob/living/carbon/human/dust_animation()
```
Why is this removed?

Ziiro replied:
Whoops. Re-added.

SamHPurp commented on code/modules/surgery/organs/tongue.dm:
```diff
@@ -190,6 +190,17 @@
+		if(2)
+			owner.apply_effect(STUTTER, 30)
+			owner.emote("scream")
+			to_chat(owner, "<span class='warning'>Alert: Vocal cords are malfunctioning.</span>")
```
no point having these two lines duplicate in each. If it'll happen regardless, stick it outside of the switch.

Ziiro replied:
You know, I could've sworn there was a reason I did it this way. But it turns out it works without the switch. fixed.

Ziiro committed:
Fixes dust on death, cleans up emote code, cleans up robotic tongue EMP code

Ziiro committed:
Icon updates

Ziiro commented:
Gibbing works. Organ icons updated.
![2017-11-08_11-49-47](https://user-images.githubusercontent.com/26494679/32582339-872cabec-c4ab-11e7-8d30-30f00b2f5f01.gif)

Cell (stomach), substance processor (liver), and robot ears.
![2017-11-08_11-52-58](https://user-images.githubusercontent.com/26494679/32582338-87136718-c4ab-11e7-909e-c30e3c8a2953.gif)

imsxz commented:
Sprites are neat. The external body parts are dropping when gibbed too, right?

Ziiro commented:
Correct. They just don't have icons at the moment.

Ziiro committed:
CHASSIS AND COLOR CODE, WORKING TOGETHER!
TWO GREAT TASTES THAT TASTE GREAT TOGETHER.
Adds the IPC-version of charcoal. "System cleaner."

Ziiro commented:
![2017-11-09_18-48-45](https://user-images.githubusercontent.com/26494679/32640797-a4413180-c587-11e7-9b11-ebebbcf863d3.gif)

I'm not sure if the solution is 100% acceptable code-wise, but it works.

Also, for future reference:
System cleaner is:
Ethanol + Chlorine + Phenol + Potassium

Ziiro commented:
Adding a lot of chassis.

Special thanks to:
Traveling Merchant for Shellguard Munitions
BRAINOS of Aurora for: Bishop Cyberkinetics 2.0, Xion Manufacturing Group 2.0, Hephaestus Industries 2.0, Zeng-Hu Pharmaceuticals
Artists for the other sets I can't credit! Because I have no idea where they came from. Bay I think? :(

Ziiro committed:
Adds lots of chassis choices. (Not actually greyscaled yet though, need to do that)
Removes ipc-mmi-posibrain code/icons for future replacement.

Ziiro committed:
Fixes random chassis, greyscales some chassis.

Ziiro committed:
Merge branch 'master' into AddingIPCs

Ziiro committed:
Thanks, Travis.

Ziiro committed:
Puts the trigger for REVIVESBYHEALING in a less stupid place

Ziiro commented:
﻿ＳＴＩＬＬ  ＡＬＩＶＥ

Can't make a commit yet because code is in progress and a mess, but:
* Brute damage can now dismember EASYDISMEMBER limbs
* IPCs have no souls to offer devils.
* There is one greyscale chassis, the rest have set colors that do not accept MUTCOLORS
* I don't think I can square the blood/oil problem right now. I have to rework the blood decal stuff for future species anyway, and it's not super critical.
* Name code is a fucking clusterfuck, and all I have to gain from messing with it for IPCs is allowing numbers in names. Which.. eh..

Ziiro committed:
...Animate me 'cause I got no soul...
WORKING POSIBRAINS, BROUGHT TO YOU BY FLATTEST!
Robotic Surgery - THANKS TO FLATTEST!
Fixes Chassis Selection/Color limiting. Only colorable chassis is the greyscale one. The rest are static.
Easily Breakable limbs with all brute damage
Balance adjustments
NOSOUL added because IPCs would benefit too much from Devil contracts
REVIVESBYHEALING now checks for hellbound

Ziiro committed:
Returns souls to toasters
also adds another hellbound check to a revivesbyhealing

SamHPurp reviewed (changes requested):
I'm not done reviewing - but I'm going to bed. Will pick up again tomorrow.

Ziiro commented:
Known issues:
* If you rejuv an IPC with missing limbs, the limbs that get added are organic. Not sure the best way to fix this. Ideally we could make rejuv proc on_species_gain again? That'd fix it. Would probably be good for other species, too.
* The noise for "spilling guts" (aka dismembering the torso) is organic and squishy. I'd like to override this if it's BODYPART_ROBOTIC, but I couldn't pinpoint where this is.
* Gibbing doesn't make noise.
* **IPCs are immune to ash storms.** However, in general they seem to be worse off than organics in lavaland for other reasons, so I'm unsure how I feel about this as a tradeoff.

FlattestGuitar commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -146,3 +146,89 @@
+			stored_mmi.loc = owner.loc
+			if(stored_mmi.brainmob)
+				var/mob/living/brain/B = stored_mmi.brainmob
+				spawn(0)
```
Because there's fancy shmancy stuff going on in the surgery code. We want to make sure we toggle it after the mind transfer and holder deletion happen properly. It didn't work reliably without the spawn.

Ziiro committed:
Fixes spacing issues, removes debug code, hopefully fixes conflict.

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,96 @@
+	male_cough_sound = 'sound/effects/mob_effects/machine_cough.ogg'
+	female_cough_sound = 'sound/effects/mob_effects/machine_cough.ogg'
+	male_sneeze_sound = 'sound/effects/mob_effects/machine_sneeze.ogg'
+	female_sneeze_sound = 'sound/effects/mob_effects/machine_sneeze.ogg'
```
Odd optimisation question for @FlattestGuitar or @AndrewMontagne ... it'd be easier for them to do the following, no?

`male_scream_sound = 'sound/effects/mob_effects/goonstation/robot_scream.ogg'`
`female_scream_sound = male_scream_sound`

Alternatively on the scream I guess you could check if female is null, and if so play male - so you don't have to worry about setting.... but I guess that relies on female inheriting as null, which it might not.... 🤔

FlattestGuitar replied:
You can't do that, because this isn't code that will run, but a class definition. What Ziiro did is optimal. Maybe you could try to fuck with a #define here, but that's overcomplicating it.

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,96 @@
+	. = ..()
+	var/obj/item/organ/appendix/appendix = C.getorganslot("appendix") // Easiest way to remove it.
+	appendix.Remove(C)
+	qdel(appendix)
```
🤔 should this `QDEL_NULL` rather than qdel?

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -239,6 +239,34 @@
+		if(ishuman(C))
+			var/mob/living/carbon/human/H = C
+			//Check if this mob's species is set and can process this type of reagent
+			var/can_process = 0
```
`FALSE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -239,6 +239,34 @@
+			var/species_handled = !(H.dna.species.handle_reagents(H, R))
+			can_process = can_process && !species_handled
+			//If the mob can't process it, remove the reagent at it's normal rate without doing any addictions, overdoses, or on_mob_life() for the reagent
+			if(can_process == 0)
```
`if(!can_process)`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
 	return 0
 
+/datum/reagents/proc/reaction_check(mob/living/M, datum/reagent/R)
+	var/can_process = 0
```
`FALSE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -239,6 +239,34 @@
+			//If we somehow avoided getting a species or reagent_tag set, we'll assume we aren't meant to process ANY reagents (CODERS: SET YOUR SPECIES AND TAG!)
+			if(H.dna && H.dna.species.reagent_tag)
+				if((R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_SYNTHETIC))		//SYNTHETIC-oriented reagents require PROCESS_SYNTHETIC
+					can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -239,6 +239,34 @@
+				if((R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_SYNTHETIC))		//SYNTHETIC-oriented reagents require PROCESS_SYNTHETIC
+					can_process = 1
+				if((R.process_flags & ORGANIC) && (H.dna.species.reagent_tag & PROCESS_ORGANIC))		//ORGANIC-oriented reagents require PROCESS_ORGANIC
+					can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -239,6 +239,34 @@
+					can_process = 1
+				//Species with PROCESS_DUO are only affected by reagents that affect both organics and synthetics, like acid and hellwater
+				if((R.process_flags & ORGANIC) && (R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_DUO))
+					can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
+		//Check if this mob's species is set and can process this type of reagent
+		if(H.dna && H.dna.species.reagent_tag)
+			if((R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_SYNTHETIC))		//SYNTHETIC-oriented reagents require PROCESS_SYNTHETIC
+				can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
+			if((R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_SYNTHETIC))		//SYNTHETIC-oriented reagents require PROCESS_SYNTHETIC
+				can_process = 1
+			if((R.process_flags & ORGANIC) && (H.dna.species.reagent_tag & PROCESS_ORGANIC))		//ORGANIC-oriented reagents require PROCESS_ORGANIC
+				can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
+				can_process = 1
+			//Species with PROCESS_DUO are only affected by reagents that affect both organics and synthetics, like acid and hellwater
+			if((R.process_flags & ORGANIC) && (R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_DUO))
+				can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
+	//We'll assume that non-human mobs lack the ability to process synthetic-oriented reagents (adjust this if we need to change that assumption)
+	else
+		if(R.process_flags != SYNTHETIC)
+			can_process = 1
```
`TRUE`

SamHPurp commented on code/modules/surgery/helpers.dm:
```diff
@@ -2,6 +2,8 @@
 		return
 
+
+
```
🤔

SamHPurp commented on code/modules/surgery/helpers.dm:
```diff
@@ -13,6 +15,11 @@
 
+	if((affecting.status == ORGAN_ROBOTIC && !istype(I, /obj/item/screwdriver)) && (affecting.status == ORGAN_ORGANIC && !I.sharpness))
+	//if it's a robotic organ and you're not using a screwdriver OR an organic organ and you're not using something sharp.
+	//shoot me, I know.
```
🔫

SamHPurp commented on code/modules/surgery/organ_manipulation.dm:
```diff
@@ -24,10 +24,12 @@
 	var/implements_mend = list(/obj/item/cautery = 100, /obj/item/weldingtool = 70, /obj/item/lighter = 45, /obj/item/match = 20)
 	var/current_type
+	var/mends_the_incision = "mends the incision in"
+	var/mend_the_incision = "mend the incision in"//so we can reuse the whole thing for robotic surgery
```
picky I know - can you make sure the order is the same in any instances of these. The order is flipped in `code/modules/surgery/organ_manip_robotic.dm` lines `32-33` for instance

SamHPurp commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -86,3 +86,21 @@
+		if(1)
+			to_chat(owner, "<span class='warning'>Alert: Your Substance Processor is nonfunctional and requires repairs or replacement.</span>")
+		if(2)
+			to_chat(owner, "<span class='warning'>Alert: Your Subtance Processor is damaged and requires repairs.</span>")
```
`Subtance` -> `Substance`

SamHPurp commented on code/modules/surgery/surgery.dm:
```diff
@@ -6,7 +6,7 @@
 	var/list/species = list(/mob/living/carbon/human)		//Acceptable Species
 	var/location = "chest"									//Surgery location
-	var/requires_organic_bodypart = 1						//Prevents you from performing an operation on robotic limbs
+	var/bodypart_types = BODYPART_ORGANIC						//Prevents you from performing an operation on robotic limbs
```
can you fix the comment spacing lol

SamHPurp commented on config/game_options.txt:
```diff
@@ -400,10 +400,11 @@ ROUNDSTART_RACES human
+#ROUNDSTART_RACES slime
 ROUNDSTART_RACES ethari
-ROUNDSTART_RACES pod
+#ROUNDSTART_RACES pod
```
Goofy that this is here, as thats what is live.

Ziiro replied:
Maybe we should move them to the section below instead of living them here just commented out 🤔

SamHPurp reviewed (dismissed):
More bits and bobs from me

Ziiro committed:
Fixes
Makes chassis selection code much nicer (not perfect)
Fixes nutrition from stable plasma (oops)
Spelling error
Whitespace errors
Useless comment removal

Ziiro committed:
Forgot one.

Ziiro committed:
Whoops! Changed wrong 0 and 1.

Ziiro committed:
Update merge, resolving conflict

Ziiro committed:
REVERT, MISTAKES WERE MADE

This reverts commit fa7d7575690b32ee6b6fa5000b855bc9ae0ce23b.

AndrewMontagne committed:
Merge branch 'master' into AddingIPCs

AndrewMontagne committed:
Merge branch 'master' into AddingIPCs

FlattestGuitar commented on code/__DEFINES/mobs.dm:
```diff
@@ -42,6 +42,15 @@
+
+// Reagent type flags, defines the types of mobs this reagent will affect
+#define ORGANIC 1
+#define SYNTHETIC 2
```
Keep these two classes and leave the rest out, there's no point. Just use `ORGANIC | SYNTHETIC` when you want to get both at once.

Ziiro replied:
`Process_DUO` has been scoured from the codebase.

It was a strange thing: `PROCESS_DUO` would only work on `ORGANIC | SYNTHETIC`. Nothing else. It wouldn't process `ORGANIC` and it wouldn't process `SYNTHETIC`. Just.. `ORGANIC | SYNTHETIC`

¯\\_(ツ)_/¯

FlattestGuitar commented on code/modules/mob/dead/new_player/sprite_accessories.dm:
```diff
@@ -60,6 +60,7 @@
 	var/dimension_x = 32
 	var/dimension_y = 32
 	var/center = FALSE	//Should we center the sprite?
+	var/fake_limbs_id //Used for features that switch out limb icons
```
A more descriptive comment on what this does please.

FlattestGuitar commented on code/modules/surgery/organ_manipulation.dm:
```diff
@@ -64,6 +66,30 @@
+		if(!affected)
+			return -1 //???
+		//I'd like to take this comment to point out that it IS possible to put a posibrain into a species that never had one
+		//This might be a terrible idea, but I think it might lead to some entertaining shenanigans
```
Having given this some thought, I think it really is a terrible idea. Can you add the check to make sure only IPCs can get posibrains inside of them? This will cause too many issues with changelings, doublebrains and other crap.

Ziiro replied:
Would simply using "isipc" in this case? 

```
if(!isipc(target))
	to_chat(user, "<span class='notice'>[target] does not have the proper connectors to interface with [tool].</span>")
	return -1
```

Ziiro replied:
Added this, Line 80.

FlattestGuitar commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -18,12 +18,12 @@
 	decoy_override = TRUE
 
-/obj/item/organ/brain/Insert(mob/living/carbon/C, special = 0)
+/obj/item/organ/brain/Insert(mob/living/carbon/C, special = 0, no_id_transfer = FALSE)
```
Was it you or me that added `no_id_transfer`? It doesn't show up anywhere else in code and I don't think it's ever called properly...

FlattestGuitar replied:
Ah, it's in the species change code. All good.

FlattestGuitar commented on code/modules/mob/living/brain/posibrain.dm:
```diff
@@ -46,6 +51,10 @@ GLOBAL_VAR(posibrain_notify_cooldown)
 		brainmob = new(src)
 	if(is_occupied())
 		to_chat(user, "<span class='warning'>This [name] is already active!</span>")
```
Should probably remove this line if we want to leave the speaker thing in here.

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,90 @@
+	for(var/X in C.bodyparts)
+		var/obj/item/bodypart/O = X
+		O.change_bodypart_status(BODYPART_ROBOTIC) // Makes all Bodyparts robotic.
+		var/chassis = C.dna.features["ipc_chassis"]
```
Correct me if I'm wrong, but the whole point of having a `/datum/sprite_accessory/ipc_chassis` datum is being able to properly assign `MUTCOLORS` and `limbs_id`, right? Why don't you use those?

FlattestGuitar commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
+			if((R.process_flags & ORGANIC) && (H.dna.species.reagent_tag & PROCESS_ORGANIC))		//ORGANIC-oriented reagents require PROCESS_ORGANIC
+				can_process = TRUE
+			//Species with PROCESS_DUO are only affected by reagents that affect both organics and synthetics, like acid and hellwater
+			if((R.process_flags & ORGANIC) && (R.process_flags & SYNTHETIC) && (H.dna.species.reagent_tag & PROCESS_DUO))
```
Should be:
`if((R.process_flags & (ORGANIC | SYNTHETIC)) && (H.dna.species.reagent_tag & (PROCESS_ORGANIC | PROCESS_SYNTHETIC))`

For reagents that work on both Synths and Organics. No species should have both flags enabled, as far as I can tell.

Also, I'd cut down the flags into just one set - `PROCESS_ORGANIC` and `PROCESS_SYNTHETIC`

FlattestGuitar commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -476,6 +504,25 @@
 		del_reagent(R.id)
 	return 0
 
+/datum/reagents/proc/reaction_check(mob/living/M, datum/reagent/R)
```
As nice as this proc is, it is never actually invoked anywhere. Please make it be used?

Ziiro replied:
Now added it in `proc/reaction`, like 538.

FlattestGuitar commented:
Off the top of my head, things we're still missing:
- Changelings being banned from being and nomming IPCs
- IPC part printing at Robotics

Ziiro committed:
Lots of fixes
-Adds render_like_organic var to limbs to pretty up code
-Fixes a chemistry true/false because I'm retarded
-Removes PROCESS_DUO
-Adds IPC Components to Mech Fab
-Adds SYNTHETICSPECIES species_trait to prevent Changeling SUCC
-Adds reaction_check to reaction in holder.dm
-Removes warning about a posibrain being active
-Removes unhelpful comments
-Some robotic organ description changes

Ziiro commented on code/game/objects/items.dm:
```diff
@@ -593,7 +593,7 @@ GLOBAL_VAR_INIT(rpg_loot_items, FALSE)
 /obj/item/proc/get_dismemberment_chance(obj/item/bodypart/affecting)
 	if(affecting.can_dismember(src))
-		if((sharpness || damtype == BURN) && w_class >= 3)
+		if((sharpness || damtype == BURN || (damtype == BRUTE && (affecting.owner.dna && affecting.owner.dna.species && (EASYDISMEMBER in affecting.owner.dna.species.species_traits)))) && w_class >= 3)
```
If anyone has suggestions for it, I'm all ears. It's a fairly specific thing to have to setup, and it's only used right here.

Ziiro commented:
Current issues:
* IPCs can be roundstart changelings.

Things I want to solve but probably won't until after release:
* No emag effect
* No way to make new bodies
* Chassis code is sub-optimal but I'm burned out on trying to do something better

Joseph-Case commented:
For the changeling issue you could probably find the code which stops mindshields from getting it, then make a PROTECT SPECIES proc or whatever its called

Ziiro committed:
blah blah blah
-IPCs can no longer be Clings.. in theory.
-Removes PROCESS_DUO I missed
-Makes IPC names consistent
-Removes unused fake_limb_id code

Ziiro commented:
Alright, I think the fixes all the outstanding issues and comments. Let me know if there's something else that needs changing or fixing.

FlattestGuitar commented on code/modules/projectiles/projectile.dm:
```diff
@@ -97,6 +97,11 @@
 				splatter_dir = get_dir(starting, target_loca)
 			if(isalien(L))
 				new /obj/effect/temp_visual/dir_setting/bloodsplatter/xenosplatter(target_loca, splatter_dir)
+			if(isipc(L)) // Not the right solution. I need it to check if the location hit is BODYPART_ROBOTIC, but I don't know how to do that.
```
var/obj/bodypart/B = L.get_bodypart(def_zone)
if(B.status == ORGAN_ROBOTIC)

or something close to that.

Ziiro committed:
BODYPART_ROBOTIC projectile interactions fixed up
Also IPC now leave oil spatters when beaten

FlattestGuitar commented on code/__DEFINES/DNA.dm:
```diff
@@ -118,4 +118,7 @@
+#define NOPAIN			30
+#define REVIVESBYHEALING 31 // Mob will revive on heal when healing past a certain point.
+#define NOSCAN			32 // Cannot be scanned by DNA scanner/Cloner scanner.
+#define SYNTHETICSPECIES	33
```
Turn that into something more specific, please. `NOCHANGELING` would be better.

FlattestGuitar commented on code/game/gamemodes/game_mode.dm:
```diff
@@ -332,6 +333,11 @@
 						candidates += player.mind				// Get a list of all the people who want to be the antagonist for this round
 
+	if(protected_species)
+		for(var/mob/dead/new_player/player in players)
```
should be cycling through `candidates`, not `players`

FlattestGuitar commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -82,7 +82,7 @@
 	if(lying && user.a_intent == INTENT_HELP)
 
-		if(I.sharpness && can_operate(src))
+		if((I.sharpness || istype(I, /obj/item/screwdriver)) && can_operate(src))//sorry for the snowflake, kids! At least the istype() check won't proc if the item is sharp to start with
```
This should also check for a coin to be used for extra ghetto surgery on toasters. It's a snowflake, but there's no better way to solve this.

FlattestGuitar commented on code/modules/mob/living/carbon/human/death.dm:
```diff
@@ -1,20 +1,35 @@
 /mob/living/carbon/human/gib_animation()
-	new /obj/effect/temp_visual/gib_animation(loc, "gibbed-h")
+	if(dna.species.species_gibs == "human")
```
All these statements should be switches.

```
switch(dna.species.species_gibs)
    if("human")
        new obj/effect...blah blah
    if("robotic")
        new obj/blah blah
...```

FlattestGuitar commented on code/modules/mob/living/carbon/human/emote.dm:
```diff
@@ -201,3 +201,102 @@
+/datum/emote/living/carbon/human/robot_tongue/clown/can_run_emote(mob/user)
+	if(!..())
+		return FALSE
+	var/obj/item/organ/tongue/T = user.getorganslot("tongue")
```
This emote's parent already checks for the robotic tongue. Just check for clown stuff here.

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,96 @@
+		var/obj/item/bodypart/O = X
+		O.change_bodypart_status(BODYPART_ROBOTIC) // Makes all Bodyparts robotic.
+		O.render_like_organic = TRUE
+		var/chassis = C.dna.features["ipc_chassis"]
```
Did what I proposed earlier not work? This really can't be implemented like this. The below switch() is not acceptable.

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,96 @@
+		H.reagents.remove_reagent(chem.id, REAGENTS_METABOLISM)
+		return 1
+
+/datum/species/ipc/spec_attacked_by(obj/item/I, mob/living/user, obj/item/bodypart/affecting, intent, mob/living/carbon/human/H)
```
That same code is repeated more generally for all limbs. Shouldn't we remove one of the ocurrences?

Ziiro replied:
Per our discord discussion: This here is specifically related to melee attacks, not ranged. I could go put it into the `attacked_by` code in `item_attack.dm` directly and attach it to `BODYPART_ROBOTIC` like I did projectile. (But last I checked, attacked_by doesn't check for hit location)

Ziiro replied:
This is for melee attacks. I could move it elsewhere, but it seemed appropriate (originally) to put it under spec_attacked_by. I'll move it somewhere else with a more generalized BODYPART_ROBOTIC check.

Ziiro committed:
The fixes never end
SYNTHETICSPECIES to NOCHANGELING
Coin operated surgery
Clean up gib code
Clean up clown emotes
Fixes game_mode protected_species shenanigans

Ziiro committed:
slightly less shitcode
-Moves the sparking/oil for melee attacks and makes it just check for bodypart_robotic
-REVIVESBYHEALING now cures husk

Ziiro commented:
Known issues:
* Chassis code still sucks (It's shitcode but I don't know how improve it. It's contained in ipc.dm at least.)
* No way to make new bodies 
* Admin Rejuving missing limbs gives them organic limbs. As a workaround, just do VV -> "Set Species" from the drop down, and set them as IPC. It'll do their `on_species_gain` again and fix the problem.

Ziiro committed:
CHASSIS CODE IS GOOD NOW
Also fixed human_parts.dmi conflict

Ziiro committed:
Merges updates and resolves conflicts

Ziiro committed:
span fixes and adds surgery

Ziiro committed:
STUFF
-Greyscale Accessories
-Removes Morpheus Cyberkinetics (Black)
-Adds Monitor Switch (Thanks Flattest)
-Fixes Amputation surgery
-Adds a new monitor screen

Ziiro committed:
Adds charging cable.

Ziiro committed:
FIXES POSIBRAIN CREATED SPAM

Ziiro committed:
Fixes Chassis code add MUTCOLORS to traits literally ten billion times

Ziiro committed:
Merge remote-tracking branch 'remotes/upstream/master' into AddingIPCs

FlattestGuitar commented:
* Can't add numbers to IPC names

SamHPurp commented:
EMPs are too weak.

I ate 4 EMP grenades on the same tile as me, I was stunned briefly, sure... but I fixed myself up entirely with less than half a roll of cable.

Imo. Give each limb a chance to pop off, increase the damage done, and if the limb doesn't pop off, make it cause sparks.

SamHPurp commented:
Few things also:

The gif action button to change your monitor seems distracting af.

Once I ate all those EMP's I had so far negative nutrition I drained 3-4 full APCs to satiate myself.

IT also fully drains those APCs so much that the lights flicker off when you're done with it. Dunno if intentional but figured I'd note it.

Ziiro commented:
I disagree vehemently with being able to kill someone using pure EMP. You should have to do *something* else as well as EMP them. EMP and toolbox. EMP and space them. EMP and _put in some effort_. 2TC should not be a free kill. Period.

This has the additional effect of EMPs not being accidental kills in the case of some member of security firing an EMP rifle, or the EMP from the singularity (but that may kill them anyway, based on the duration of the stun)

Ziiro commented:
* Slimes can do Cellular damage to IPCs
* IPCs drain values on APCs is too damn much (maybe related to below)
* EMP draining nutrition can get the value **lower than 0**, which causes problems to fix. (Probably should be encouraged at this point to use plasma, though, because of the unit/nutrition exchange rate) But this is a fairly easy fix

>IPCs have tastebuds

I need to test this, but I'm not sure the tags of all food being "Toxic" would work if they *didn't* taste anything.

Joseph-Case commented:
Imo, EMPing an IPC should drop off all their limbs except the head, so they can scream for help, then still be able to be toolboxed to desth

EldritchSigma commented:
IPCs are indeed powersinks right now. I made the lights shut off on the mining outpost by just eating power

Ziiro committed:
FEEDBACK FIXES
-IPCs can have numbers in names
-IPC Eating adjustments
-Limbs have a 20% chance to pop off when EMP'd.
-General EMP adjustments
-IPCs no longer take clone damage
-IPCs now take some toxic damage
-Having no taste sensitivity literally means you taste nothing, now
-Monitor graphic less distracting

Ziiro committed:
Screen tweaks.
-Certain screens are now RGB-disabled
-Screen change function can now change colors as well
-Adds a few more screens (Thanks, Brasnak!)
-IPC Screens now turn off on death

Ziiro committed:
ADDS ASS_PIC FOR IPCS
very important!!!!

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,156 @@
+/datum/species/ipc/handle_chemicals(datum/reagent/chem, mob/living/carbon/human/H)
+	if(chem.id in list("plasma", "stable_plasma")) // IPCs have plasma batteries.
+		H.nutrition += 5
+		if(H.nutrition > NUTRITION_LEVEL_FULL)
```
`H.nutrition = min(H.nutrition + 5, NUTRITION_LEVEL_FULL)`

FlattestGuitar commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,156 @@
+			to_chat(H, "<span class='warning'>\The [A] has no more charge.</span>")
+			break
+		A.charging = 1
+		if(A.cell.charge >= 250)
```
Change this check to `A.cell.charge >= 500` so we don't power off APCs.

FlattestGuitar commented on code/modules/mob/living/taste.dm:
```diff
@@ -25,7 +25,7 @@
 			"defeat","pain","bliss","revenge","poison","time","space","death","life","truth","lies","justice","memory",\
 			"regrets","your soul","suffering","music","noise","blood","hunger","the american way")
-		if(text_output != last_taste_text || last_taste_time + 100 < world.time)
+		if((text_output != last_taste_text || last_taste_time + 100 < world.time) && taste_sensitivity < 101) // 101 means literally no taste ability.
```
Change that 101 to a define. I'd make that -1. `#define TASTE_SENSITIVITY_NO_TASTE -1`
Then add that to all species and make a `==` check here.

FlattestGuitar commented on code/modules/projectiles/projectile.dm:
```diff
@@ -97,6 +97,14 @@
 			if(isalien(L))
 				new /obj/effect/temp_visual/dir_setting/bloodsplatter/xenosplatter(target_loca, splatter_dir)
+			var/obj/item/bodypart/B = L.get_bodypart(def_zone)
+			if(B.status == BODYPART_ROBOTIC) // So if you hit a robotic, it sparks instead of bloodspatters
```
Honestly, this whole thing should be a bodypart proc.

`/obj/item/bodypart/proc/make_blood()`

Then redefined for aliums and beep boops.
Will also work with bullet hits, I'm pretty sure.

Ziiro replied:
I'm not sure I agree:

Projectile code has to do a few calculations to check where the bloodspatter is coming from/going, as well as the location to drop the blood spatter. It also takes aliens into account. (obviously.)

Melee is just a chance to drop splatter when getting hit by sharp weapons.

Offloading just the bloodsplatter/oilsplatter into a proc seems like it's just moving code around without any particular gains, but maybe I'm just not understanding what you want as a part of it?

FlattestGuitar commented on code/modules/reagents/chemistry/holder.dm:
```diff
@@ -494,6 +535,10 @@
+				var/check = reaction_check(A, R)
+				if(!check)
+					continue
+				R.reaction_mob(A, method, R.volume * volume_modifier)
```
Riddle me this: You add a `reaction_mob()` call here, but this isn't a proc you added in this PR. Does that mean there's another `reaction_mob()` being called somewhere else? Seems like a bad idea to me.

Ziiro replied:
I guess it is, in theory, being called twice on touch then, isn't it? Hmm.

I should probably remove this then.

FlattestGuitar commented on code/modules/reagents/chemistry/recipes/medicine.dm:
```diff
@@ -38,6 +38,12 @@
+/datum/chemical_reaction/system_cleaner
+	name = "System Cleaner"
+	id = "system_cleaner"
+	results = list("system_cleaner" = 2)
```
I'd make it 4u for the consistency.

Ziiro committed:
taste and other minor tweaks
-taste
-IPC APC draining numbers changed so IPCs can't depower an area
-IPC drain verbage fixes
-system cleaner made consistent

Ziiro committed:
removes redundant reaction_mob()

Ziiro committed:
IPCs will no longer get stuck in crit limbo

Ziiro committed:
shitcode approaching critical mass. refactor imminent
-Rejuving IPCs now works
-Fixes some splint code runtiming due to NOBONES

Ziiro committed:
We're getting there
Fixes rejuv/chassis stuff
Fixes IPC facial display for masks/helmets
Fixes redundancy on a reagent
Removes Androids and changes Android Slimetoxin

SamHPurp commented:
![image](https://user-images.githubusercontent.com/7415122/34074140-d6cc12d8-e2a0-11e7-846a-36ba5ee9f37b.png)
My neck is invisible yo

Ziiro commented:
current issues:

* A few sprite issues (the one above as well as hands with gloves are strange)
* IPCs can't speak a language, only understand :6
* suicide messages are all organic ones

SamHPurp commented:
`You feel kind of iffy...`

`You feel kind of iffy...`

`You try to throw up, but there's nothing in your stomach!`

`You try to throw up, but there's nothing in your stomach!`

------

After eating... stunned me, put me on the floor, for aaaaages

SamHPurp commented:
If you select a name with numbers in it - the numbers aren't there in the final version.

Ziiro commented:
>After eating... stunned me, put me on the floor, for aaaaages

Don't jam food into your computer

[NOT AN ISSUE]
(for real, it's known behavior. I find it hilarious. Don't eat. Seriously. It fucks you up.)

>If you select a name with numbers in it - the numbers aren't there in the final version.

Not sure I follow - I think there's a bug where name numbers aren't saved, so if you join a new round and you previously had a name with numbers in it, it doesn't work.

SamHPurp commented:
>(for real, it's known behavior. I find it hilarious. Don't eat. Seriously. It fucks you up.)

Not sure I like this behaviour. Too easy to feed someone something if they aren't paying attention for the split second to takes to force feed someone something (particularly clowns tripping people?) and have them be stunned for enormous amounts of time.

SamHPurp commented:
@Ziiro regarding numbers. If you hit "generate random name" you can get numbers in your name.

However every time Spark tries to join as `Spark 5.5`, we end up having to rename him because he comes `Spark . `

Ziiro committed:
Second round of feedback fixes
-IPCs can no longer eat
-Numbers in names should now persist in save/load
-Some sprite fixes
-Slimetoxins now work on IPCs
-Clearer verbage in liver damage lines
-On spawn message about some basics of IPC mechanics

Ziiro committed:
Misc Improvements
-Improves mouth code (thanks, Flattest)
-Makes it so you can't attach organic limbs to robot bodies. Whoops!
-Adds a few more screens (thanks Brasnak!)
-A few sprite improvements

SamHPurp commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -101,6 +101,8 @@
 	switch(severity)
 		if(1)
-			to_chat(owner, "<span class='warning'>Alert: Your Substance Processor is nonfunctional and requires replacement.</span>")
+			to_chat(owner, "<span class='warning'>Alert: Your Substance Procesor has been severely damaged. Uncontained chemicals are affecting performance.</span>")
```
typo `Procesor `

SamHPurp commented on code/modules/surgery/organs/liver.dm:
```diff
@@ -101,6 +101,8 @@
 		if(2)
-			to_chat(owner, "<span class='warning'>Alert: Your Substance Processor is damaged.</span>")
\ No newline at end of file
+			to_chat(owner, "<span class='warning'>Alert: Your Substance Processor is damaged:Uncontained chemicals are affecting performance.</span>")
```
space missing.

Also I'm not a fan of the language `uncontained chemicals`

SamHPurp commented on code/modules/surgery/bodyparts/bodyparts.dm:
```diff
@@ -65,18 +65,19 @@
-				user.temporarilyRemoveItemFromInventory(src, TRUE)
-				attach_limb(C)
-				return
+			if((src.status == BODYPART_ORGANIC && !(ROBOTIC_LIMBS in H.dna.species.species_traits))) // no meat limbs in robotic sockets, thanks.
```
`if((src.status`

src is implied, no?

SamHPurp commented on code/modules/reagents/reagent_containers.dm:
```diff
@@ -73,6 +69,12 @@
 		return 0
+	if(!eater.has_mouth())
+		if(eater == user)
+			to_chat(world, "<span class='warning'>How can you hope to eat if you have no mouth, silly?</span>")
```
to chat, world? :o 

Also not a fan of the language here.

SamHPurp commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -281,15 +281,22 @@
 			B.render_like_organic = TRUE
 
+	if(NOMOUTH in species_traits)
+		for(var/obj/item/bodypart/head/head in C.bodyparts)
```
`head/head`? can we just use `head/H` if its a temporary variable

SamHPurp commented on code/game/gamemodes/changeling/changeling.dm:
```diff
@@ -18,6 +18,7 @@ GLOBAL_LIST_INIT(slot2type, list("head" = /obj/item/clothing/head/changeling, "w
 	antag_flag = ROLE_CHANGELING
 	restricted_jobs = list("AI", "Cyborg")
 	protected_jobs = list("Security Officer", "Warden", "Detective", "Head of Security", "Captain", "Blueshield", "Brig Physician")
+	protected_species = list("ipc")
```
❄️

SamHPurp replied:
Can we not just check `NOCHANGELING`

Ziiro replied:
Pretty sure, no. Because that involves checking species traits of a species they haven't spawned in as yet. We have to check for their species choice. Look at the code for protected_species in game_modes

FlattestGuitar replied:
What Ziiro said. No solid workaround.

We could still tie the whole `NOCHANGELING` thing to `NOSCAN` though to eliminate the define.

Ziiro replied:
NOSCAN could also be used for Vox/uncloneable races, which can be clings/clinged

SamHPurp commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -146,3 +146,81 @@
+				var/mob/living/brain/B = stored_mmi.brainmob
+				spawn(0)
+					if(B)
+						B.stat = 0//why the fuck this works I have no idea, but it actually does
```
👀

SamHPurp commented on code/modules/mob/living/brain/brain_item.dm:
```diff
@@ -146,3 +146,81 @@
+	icon = stored_mmi.icon
+	icon_state = stored_mmi.icon_state
+
+/obj/item/organ/brain/mmi_holder/posibrain/New(var/obj/item/device/mmi/MMI)
```
Can we not use `Initialize` and `. = ..()` instead?

FlattestGuitar replied:
Shit, we should.

I'll look at that when I have some time.

SamHPurp commented on code/modules/mob/living/brain/say.dm:
```diff
@@ -1,6 +1,10 @@
+	var/obj/item/device/mmi/posibrain/P = container
 	if(!(container && istype(container, /obj/item/device/mmi)))
 		return //No MMI, can't speak, bucko./N
+	if(P && P.silenced == TRUE)
```
the `== TRUE` is superfluous

SamHPurp commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -162,6 +162,13 @@
+		var/datum/effect_system/spark_spread/spark_system = new /datum/effect_system/spark_spread
+		spark_system.set_up(1, 0, src)
+		spark_system.attach(src)
+		spark_system.start()
```
whoaaaaa gross - no no no.

You're building and setting up a new thing each time!!!!

Create it once during `Initialize` then start it when you need to trigger it.

*scream *scream *scream

SamHPurp commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -162,6 +162,13 @@
+		spark_system.attach(src)
+		spark_system.start()
+		if(prob(25))
+			new /obj/effect/decal/cleanable/oil(src.loc)
```
pretty sure `src.` is superfluous

Ziiro replied:
It is not. Removed and tested and it stops working.

SamHPurp replied:
alright cool, I wasn't certain on this one. I assume it is getting confused and trying to take the implication from the new item.

SamHPurp commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -475,6 +482,9 @@
 					Stun(100)
+			if((EASYDISMEMBER in L.owner.dna.species.species_traits) && L.body_zone != "chest")
+				if(prob(20))
+					L.dismember(BRUTE)
```
Are these limbs still taking damage?

Also, if its a human with a mechanical heart.... does it try to dismember your torso?

Ziiro replied:
Yes. They take damage before this is checked

A human with a mechanical heart has no relevance: They do not have easy dismember. Additionally that's ORGAN_ROBOTIC, not BODYPART_ROBOTIC (all this applies to)

Additionally this ignores the torso, because getting EMP'd and spilling your posibrain for instant death is silly.

SamHPurp commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -55,6 +57,9 @@
 	var/fixed_mut_color = "" //to use MUTCOLOR with a fixed color that's independent of dna.feature["mcolor"]
+	var/reagent_tag = PROCESS_ORGANIC //Used for metabolizing reagents. We're going to assume you're a meatbag unless you say otherwise.
+	var/species_gibs = "human"
+	var/allow_numbers_in_name // Can this species use numbers in it's name?
```
❄️ ;)

Ziiro replied:
`comment entirely unnecessary`

FlattestGuitar replied:
a) this is the opposite of snowflaking
b) its*

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,161 @@
+	clonemod = 0
+	siemens_coeff = 1.5 // Overload!
+	species_traits = list(NOBREATH, NOBLOOD, RADIMMUNE, VIRUSIMMUNE, NOZOMBIE, EASYDISMEMBER, EASYLIMBATTACHMENT, NOPAIN, NO_BONES, NOTRANSSTING, MUTCOLORS, REVIVESBYHEALING, NOSCAN, NOCHANGELING, NOHUSK, ROBOTIC_LIMBS, NOMOUTH)
+	mutant_organs = list(/obj/item/organ/cyberimp/arm/power_cord/)
```
shouldn't have a trailing slash

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,161 @@
+	if(chem.id in list("plasma", "stable_plasma")) // IPCs have plasma batteries.
+		H.nutrition = min(H.nutrition + 5, NUTRITION_LEVEL_FULL)
+		H.reagents.remove_reagent(chem.id, REAGENTS_METABOLISM)
+		return 1
```
Is this still needed?

Ziiro replied:
Yes. If IPCs need nutrition in places they don't have access to an APC (Such as IPCs in prison.) It can be supplied via injection or something.

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,161 @@
+	var/mob/living/carbon/human/H = user
+	var/obj/item/organ/stomach/cell/cell = H.internal_organs_slot["stomach"]
+	if(!cell)
+		to_chat(H, "<span class='warning'>You try to siphon energy from the [A], but your powercell is gone!</span>")
```
`powercell` -> `power cell`

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,161 @@
+	if(A.emagged || A.stat & BROKEN)
+		var/datum/effect_system/spark_spread/s = new /datum/effect_system/spark_spread
+		s.set_up(3, 1, A)
+		s.start()
```
*scream again

SamHPurp commented on code/modules/mob/living/carbon/human/species_types/ipc.dm:
```diff
@@ -0,0 +1,161 @@
+		H.adjustFireLoss(6) // After bodypart_robotic resistance this is ~2/second
+		if(prob(5))
+			to_chat(H, "<span class='warning'>Alert: Internal temperature regulation systems offline; thermal damage sustained. Shutdown imminent.</span>")
+			H.visible_message("[H]'s cooling system fans stutter and stall. There is a faint, rapid beeping coming from inside their chassis.")
```
`faint, rapid`

`faint, yet rapid`, `faint, but rapid` etc

SamHPurp commented on code/modules/mob/living/taste.dm:
```diff
@@ -25,11 +23,9 @@
 			"defeat","pain","bliss","revenge","poison","time","space","death","life","truth","lies","justice","memory",\
 			"regrets","your soul","suffering","music","noise","blood","hunger","the american way")
-		if(text_output != last_taste_text || last_taste_time + 100 < world.time)
+		if((text_output != last_taste_text || last_taste_time + 100 < world.time) && !NO_TASTE_SENSITIVITY) // If you have no taste sensitivity, you get no message, because you can't taste it.
```
comment entirely unnecessary, it explains something that is entirely evidence from the code.

SamHPurp commented on code/modules/power/apc.dm:
```diff
@@ -632,6 +632,8 @@
 			update_icon()
 		return
+	else if(istype(W, /obj/item/apc_powercord))
+		return//because we put our fancy code in the right places, and this is all in the powercord's afterattack() >:C
```
del comment yo

SamHPurp commented on code/modules/projectiles/projectile.dm:
```diff
@@ -97,6 +97,14 @@
+				var/datum/effect_system/spark_spread/spark_system = new /datum/effect_system/spark_spread
+				spark_system.set_up(1, 0, target_loca)
+				spark_system.attach(target_loca)
+				spark_system.start()
```
again *scream

SamHPurp commented on code/modules/surgery/bodyparts/dismemberment.dm:
```diff
@@ -58,7 +58,8 @@
 	C.add_splatter_floor(T)
-	playsound(get_turf(C), 'sound/misc/splort.ogg', 80, 1)
+	if(status != BODYPART_ROBOTIC)
+		playsound(get_turf(C), 'sound/misc/splort.ogg', 80, 1)
```
Wish we had a better audio for mechanical limbs, but oh well.

Ziiro replied:
Could just the robot gib sound

FlattestGuitar replied:
Would be better to get an original one; it's fine to leave this as is if you can't find anything pretty.

SamHPurp commented on code/modules/surgery/generic_steps.dm:
```diff
@@ -116,3 +116,50 @@
+	time = 20
+
+/datum/surgery_step/unscrew/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	user.visible_message("[user] begins to unscrew the cover panel on [target]'s [parse_zone(target_zone)].",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/generic_steps.dm:
```diff
@@ -116,3 +116,50 @@
+	time = 30
+
+/datum/surgery_step/pry_off/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	user.visible_message("[user] begins to pry open the cover panel on [target]'s [parse_zone(target_zone)].",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/generic_steps.dm:
```diff
@@ -116,3 +116,50 @@
+	time = 30
+
+/datum/surgery_step/close_hatch/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	user.visible_message("[user] begins to put the cover panel on [target]'s [parse_zone(target_zone)] back in place.",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/generic_steps.dm:
```diff
@@ -116,3 +116,50 @@
+
+/datum/surgery_step/robotic_amputation/preop(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	if(!istype(tool, /obj/item/device/multitool))
+		user.visible_message("[user] begins to cut through the circuitry in [target]'s [parse_zone(target_zone)]!", "<span class='notice'>You begin to cut through the circuitry in [target]'s [parse_zone(target_zone)]...</span>")
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/generic_steps.dm:
```diff
@@ -116,3 +116,50 @@
+		user.visible_message("[user] begins to cut through the circuitry in [target]'s [parse_zone(target_zone)]!", "<span class='notice'>You begin to cut through the circuitry in [target]'s [parse_zone(target_zone)]...</span>")
+	else
+		var/pro = pick("neatly", "calmly", "professionally", "carefully", "swiftly", "proficiently")
+		user.visible_message("[user] begins to [pro] disconnect [target]'s [parse_zone(target_zone)]!", "<span class='notice'>You begin to [pro] disconnect [target]'s [parse_zone(target_zone)]...</span>")
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/generic_steps.dm:
```diff
@@ -116,3 +116,50 @@
+
+/datum/surgery_step/robotic_amputation/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	var/mob/living/carbon/human/L = target
+	user.visible_message("[user] removes [L]'s [parse_zone(target_zone)]!", "<span class='notice'>You remove [L]'s [parse_zone(target_zone)].</span>")
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/organ_manipulation.dm:
```diff
@@ -64,6 +66,31 @@
+		if(!P.brainmob || !P.brainmob.client)
+			to_chat(user, "<span class='notice'>[tool] has no life in it, this would be pointless!</span>")
+			return -1
+		user.visible_message("[user] begins to insert [tool] into [target]'s [parse_zone(target_zone)].",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/organ_manipulation.dm:
```diff
@@ -97,20 +124,35 @@
 		current_type = "mend"
-		user.visible_message("[user] begins to mend the incision in [target]'s [parse_zone(target_zone)].",
-			"<span class='notice'>You begin to mend the incision in [target]'s [parse_zone(target_zone)]...</span>")
+		user.visible_message("[user] begins to [mend_the_incision] [target]'s [parse_zone(target_zone)].",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/organ_manipulation.dm:
```diff
@@ -97,20 +124,35 @@
 	if(current_type == "mend")
-		user.visible_message("[user] mends the incision in [target]'s [parse_zone(target_zone)].",
-			"<span class='notice'>You mend the incision in [target]'s [parse_zone(target_zone)].</span>")
+		user.visible_message("[user] [mends_the_incision] [target]'s [parse_zone(target_zone)].",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/organ_manipulation.dm:
```diff
@@ -97,20 +124,35 @@
+		spawn(1)
+			I = new /obj/item/organ/brain/mmi_holder/posibrain(tool)
+			I.Insert(target)
+			user.visible_message("[user] inserts [tool] into [target]'s [parse_zone(target_zone)]!",
```
All visible messages should be spanned up, yo

SamHPurp commented on code/modules/surgery/organs/stomach.dm:
```diff
@@ -64,3 +64,23 @@
+	zone = "chest"
+	slot = "stomach"
+	attack_verb = list("assault and battery'd")
+	desc = "A Micro-cell, for IPC use only. Do not swallow."
```
`Micro` -> `micro`

Ziiro committed:
code cleanup

Ziiro commented:
Not going to bother fixing conflicts anymore until it's 100% done

Ziiro committed:
span fixes and cleanup

FlattestGuitar commented on code/modules/power/apc.dm:
```diff
@@ -633,7 +633,7 @@
 		return
 	else if(istype(W, /obj/item/apc_powercord))
-		return//because we put our fancy code in the right places, and this is all in the powercord's afterattack() >:C
+		return
```
Please leave the comment there. Can be less inventive, but the situation should be explained.

Ziiro committed:
words

Ziiro committed:
Posibrain fixes, spark fixes, runtime fixes, limb fixes

Ziiro committed:
Merge branch 'master' into AddingIPCs

# Conflicts:
#	code/modules/mob/living/carbon/carbon.dm

FlattestGuitar commented on code/modules/mob/living/carbon/human/human_defense.dm:
```diff
@@ -162,6 +162,12 @@
+		if(I.force && I.damtype != STAMINA && affecting.status == BODYPART_ROBOTIC) // Bodpart_robotic sparks when hit, but only when it does real damage
+			if((I.force >= 5) || I.is_sharp())
+				do_sparks(1, FALSE, loc)
+			if(prob(25))
```
Move the oil splatters under the above check, so they don't trigger below 5 force either.

Also, I'd remove that `is_sharp()` check, since if something is sharp and deals less than 5 damage it's probably shit anyway

Ziiro committed:
oil splotch and spark adjustment

Ziiro committed:
Chassis name tweak

SamHPurp reviewed (approved):
Lets make a mistake

Ziiro commented:
**_*SCREAM_**

Ziiro commented:
For historical purposes:
Oversight:
https://github.com/OracleStation/OracleStation/pull/413
